File: tools/body-regions/curate-torso-mask.mjs
/**
 * Curate torso regions onto authoring RGB mask using PRODUCTION GLB UVs.
 * (Authoring .blend mesh UVs currently diverge from public/models/production GLB.)
 *
 *   node tools/body-regions/curate-torso-mask.mjs
 */
import { copyFileSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import sharp from "sharp";
import { loadMeshData } from "../body-mask/glb.mjs";

const ROOT = path.resolve(import.meta.dirname, "..", "..");
const GLB = path.join(ROOT, "public/models/production/neutro_body_v1.glb");
const PALETTE = path.join(ROOT, "assets/body-regions/neutro_body_v1_region_palette.json");
const LANDMARKS = path.join(ROOT, "assets/body-regions/neutro_body_v1_landmarks.json");
const AUTHORING = path.join(
  ROOT,
  "assets/body-regions/neutro_body_v1_anatomical_regions_authoring.png",
);
const SEED = path.join(
  ROOT,
  "public/models/interaction/neutro_body_v1_public_region_mask.png",
);
const BACKUPS = path.join(ROOT, "assets/body-regions/backups");

const TORSO = [
  "right_pectoral_region",
  "left_pectoral_region",
  "full_abdomen_region",
  "right_ribs_region",
  "left_ribs_region",
  "upper_back_region",
  "lower_back_region",
];

function parseHex(hex) {
  const h = hex.replace("#", "");
  return [
    parseInt(h.slice(0, 2), 16),
    parseInt(h.slice(2, 4), 16),
    parseInt(h.slice(4, 6), 16),
  ];
}

function axisZAt(y, samples) {
  if (!samples?.length) return -0.08;
  const ys = samples.map((s) => s.y);
  const zs = samples.map((s) => s.z);
  if (y <= ys[0]) return zs[0];
  if (y >= ys.at(-1)) return zs.at(-1);
  for (let i = 0; i < ys.length - 1; i++) {
    if (y >= ys[i] && y <= ys[i + 1]) {
      const t = (y - ys[i]) / Math.max(1e-6, ys[i + 1] - ys[i]);
      return zs[i] * (1 - t) + zs[i + 1] * t;
    }
  }
  return zs.at(-1);
}

function classify(x, y, z, levels, samples) {
  const axz = axisZAt(y, samples);
  const th = (Math.atan2(x, z - axz) * 180) / Math.PI;
  const ath = Math.abs(th);
  const { neckBase: neck, infraclavicular: clav, inframammary: imf, inferiorScapular: scap, waist, iliacCrest: iliac } = levels;

  if (y > neck + 0.02 || y < iliac - 0.02) return null;
  if (Math.abs(x) > 0.28 && ath > 55 && y > clav - 0.08) return null;

  const latN = Math.min(1, Math.abs(x) / 0.14);
  const imfCurve = imf - 0.008 + 0.045 * latN ** 1.55;
  const topPec = clav + 0.008;
  const front = z - axz;

  // Pectorals — full mound to IMF, sternum, axilla
  if (y >= imfCurve - 0.012 && y <= topPec && ath <= 75 && front > -0.015) {
    const t = Math.max(0, Math.min(1, (y - imfCurve) / Math.max(0.05, topPec - imfCurve)));
    const maxAth = 48 + 22 * t;
    if (ath <= maxAth && front > -0.008) {
      if (x < -0.003) return "right_pectoral_region";
      if (x > 0.003) return "left_pectoral_region";
      return x <= 0 ? "right_pectoral_region" : "left_pectoral_region";
    }
  }

  // Abdomen
  let abTop = imfCurve - 0.01;
  const abTopC = imf - 0.02;
  if (ath < 32) abTop = abTopC + (abTop - abTopC) * (ath / 32);
  if (y >= iliac + 0.048 && y <= abTop && ath <= 60 && front > 0.0) {
    return "full_abdomen_region";
  }

  // Ribs wrap
  if (y >= waist - 0.04 && y <= clav - 0.03 && ath >= 46 && ath <= 138) {
    if (ath >= 54 || y < imfCurve - 0.02) {
      return x < 0 ? "right_ribs_region" : x > 0 ? "left_ribs_region" : null;
    }
  }

  // Upper back
  if (y >= scap - 0.015 && y <= neck - 0.008 && ath >= 92 && Math.abs(x) < 0.23) {
    return "upper_back_region";
  }

  // Lower back
  if (y >= iliac + 0.03 && y <= scap + 0.005 && ath >= 88 && Math.abs(x) < 0.21) {
    return "lower_back_region";
  }

  return null;
}

function colorNear(rgb, target, tol = 10) {
  return (
    Math.abs(rgb[0] - target[0]) <= tol &&
    Math.abs(rgb[1] - target[1]) <= tol &&
    Math.abs(rgb[2] - target[2]) <= tol
  );
}

async function main() {
  mkdirSync(BACKUPS, { recursive: true });
  const stamp = new Date().toISOString().replace(/[:.]/g, "-");
  if (AUTHORING) {
    try {
      copyFileSync(AUTHORING, path.join(BACKUPS, `pre_glb_curate_${stamp}.png`));
    } catch {}
  }

  const palette = JSON.parse(readFileSync(PALETTE, "utf8"));
  const lm = JSON.parse(readFileSync(LANDMARKS, "utf8"));
  const mesh = loadMeshData(GLB);
  const size = palette.resolution || 4096;

  // Start from seed indexed → keep limbs/head, or from authoring if present
  let basePath = AUTHORING;
  try {
    readFileSync(AUTHORING);
  } catch {
    basePath = SEED;
  }

  // Build RGB canvas from seed indexed mask for non-torso preservation
  const seed = await sharp(SEED).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
  const byIndex = new Map([[0, [0, 0, 0]]]);
  for (const [id, e] of Object.entries(palette.regions)) {
    byIndex.set(e.runtimeIndex, parseHex(e.authoringColor));
  }
  const torsoColors = TORSO.map((id) => parseHex(palette.regions[id].authoringColor));
  const out = Buffer.alloc(size * size * 3);

  for (let i = 0; i < size * size; i++) {
    const idx = seed.data[i * 4];
    let rgb = byIndex.get(idx) || [0, 0, 0];
    // clear prior torso colors
    if (torsoColors.some((c) => colorNear(rgb, c, 8))) rgb = [0, 0, 0];
    out[i * 3] = rgb[0];
    out[i * 3 + 1] = rgb[1];
    out[i * 3 + 2] = rgb[2];
  }

  const colorOf = Object.fromEntries(
    TORSO.map((id) => [id, parseHex(palette.regions[id].authoringColor)]),
  );

  const STEPS = 8;
  let painted = 0;
  const P = mesh.positions;
  const UV = mesh.uvs;
  const I = mesh.indices;

  for (let t = 0; t < mesh.triangleCount; t++) {
    const i0 = I[t * 3];
    const i1 = I[t * 3 + 1];
    const i2 = I[t * 3 + 2];
    for (let a = 0; a <= STEPS; a++) {
      for (let b = 0; b <= STEPS - a; b++) {
        const wa = a / STEPS;
        const wb = b / STEPS;
        const wc = 1 - wa - wb;
        const x = P[i0 * 3] * wc + P[i1 * 3] * wa + P[i2 * 3] * wb;
        const y = P[i0 * 3 + 1] * wc + P[i1 * 3 + 1] * wa + P[i2 * 3 + 1] * wb;
        const z = P[i0 * 3 + 2] * wc + P[i1 * 3 + 2] * wa + P[i2 * 3 + 2] * wb;
        const rid = classify(x, y, z, lm.levels, lm.axisZSamples);
        if (!rid) continue;
        const u = UV[i0 * 2] * wc + UV[i1 * 2] * wa + UV[i2 * 2] * wb;
        const v = UV[i0 * 2 + 1] * wc + UV[i1 * 2 + 1] * wa + UV[i2 * 2 + 1] * wb;
        const px = Math.min(size - 1, Math.max(0, Math.round(u * (size - 1))));
        const py = Math.min(size - 1, Math.max(0, Math.round((1 - v) * (size - 1))));
        const col = colorOf[rid];
        for (let dy = -1; dy <= 1; dy++) {
          for (let dx = -1; dx <= 1; dx++) {
            const xx = Math.min(size - 1, Math.max(0, px + dx));
            const yy = Math.min(size - 1, Math.max(0, py + dy));
            const o = (yy * size + xx) * 3;
            out[o] = col[0];
            out[o + 1] = col[1];
            out[o + 2] = col[2];
            painted++;
          }
        }
      }
    }
  }

  await sharp(out, { raw: { width: size, height: size, channels: 3 } })
    .png({ compressionLevel: 9 })
    .toFile(AUTHORING);

  console.log("WROTE", path.relative(ROOT, AUTHORING));
  console.log("PAINTED_STAMPS", painted);
  console.log("CURATE_TORSO_GLB_OK");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: tools/body-regions/blender/curate_torso_mask.py
"""
Curate torso anatomical regions onto the authoring UV mask.

Uses landmark-guided surface classification + dense UV rasterization so
boundaries follow painted anatomy rather than low-poly face edges.

Does NOT wipe limbs/head/pelvis — only rewrites torso gate IDs.

Usage:
  blender.exe authoring.blend --background --python tools/body-regions/blender/curate_torso_mask.py
"""

from __future__ import annotations

import json
import math
import shutil
import sys
from datetime import datetime
from pathlib import Path

import bpy
from mathutils import Vector

ROOT = Path(__file__).resolve().parents[3]
PALETTE = ROOT / "assets/body-regions/neutro_body_v1_region_palette.json"
LANDMARKS = ROOT / "assets/body-regions/neutro_body_v1_landmarks.json"
AUTHORING = ROOT / "assets/body-regions/neutro_body_v1_anatomical_regions_authoring.png"
BACKUPS = ROOT / "assets/body-regions/backups"
PAINTABLE = "NEUTRO_BODY_MASK_AUTHORING"
IMAGE_NAME = "AnatomicalRegionsAuthoring"

TORSO_IDS = [
    "right_pectoral_region",
    "left_pectoral_region",
    "full_abdomen_region",
    "right_ribs_region",
    "left_ribs_region",
    "upper_back_region",
    "lower_back_region",
]


def hex_rgb(hex_color: str):
    h = hex_color.lstrip("#")
    return (
        int(h[0:2], 16) / 255.0,
        int(h[2:4], 16) / 255.0,
        int(h[4:6], 16) / 255.0,
        1.0,
    )


def find_body():
    obj = bpy.data.objects.get(PAINTABLE)
    if obj and obj.type == "MESH":
        return obj
    best = None
    n = -1
    for o in bpy.data.objects:
        if o.type == "MESH" and len(o.data.vertices) > n:
            best = o
            n = len(o.data.vertices)
    return best


def load_image():
    img = bpy.data.images.get(IMAGE_NAME)
    if img is None:
        img = bpy.data.images.load(str(AUTHORING))
        img.name = IMAGE_NAME
    else:
        img.filepath = str(AUTHORING)
        try:
            img.reload()
        except Exception:
            pass
    return img


def detect_runtime_to_blender(obj):
    """Return (to_blender, to_runtime, mode).

    Landmarks are runtime Y-up Z-front.
    Authoring blend is Z-up: runtime (x,y,z) → blender (x, z, y).
    """
    bb = [obj.matrix_world @ Vector(c) for c in obj.bound_box]
    dy = max(v.y for v in bb) - min(v.y for v in bb)
    dz = max(v.z for v in bb) - min(v.z for v in bb)
    if dz >= dy * 0.95:
        def to_blender(p):
            return Vector((p[0], p[2], p[1]))

        def to_runtime(p):
            return (p.x, p.z, p.y)

        return to_blender, to_runtime, "z_up_xz_swap"
    def to_blender(p):
        return Vector((p[0], p[1], p[2]))

    def to_runtime(p):
        return (p.x, p.y, p.z)

    return to_blender, to_runtime, "y_up"


def axis_z_at(y, samples):
    if not samples:
        return -0.08
    ys = [s["y"] for s in samples]
    zs = [s["z"] for s in samples]
    if y <= ys[0]:
        return zs[0]
    if y >= ys[-1]:
        return zs[-1]
    for i in range(len(ys) - 1):
        if ys[i] <= y <= ys[i + 1]:
            t = (y - ys[i]) / max(1e-6, ys[i + 1] - ys[i])
            return zs[i] * (1 - t) + zs[i + 1] * t
    return zs[-1]


def theta_deg(p_runtime, axis_z):
    return math.degrees(math.atan2(p_runtime[0], p_runtime[2] - axis_z))


def classify_torso(p_rt, lm, levels, axis_samples):
    """
    Landmark-guided torso classification in runtime coords.
    Returns region id or None (leave existing non-torso).
    """
    x, y, z = p_rt
    axz = axis_z_at(y, axis_samples)
    th = theta_deg(p_rt, axz)
    ath = abs(th)

    neck = levels["neckBase"]
    clav = levels["infraclavicular"]
    apex = levels["breastApex"]
    imf = levels["inframammary"]
    scap = levels["inferiorScapular"]
    waist = levels["waist"]
    iliac = levels["iliacCrest"]

    # Outside torso vertical band → not our rewrite target
    if y > neck + 0.02 or y < iliac - 0.02:
        return None

    # Arms roughly: far lateral and not on torso cylinder — leave alone if clearly limbs
    # Soft torso width gate
    torso_half = 0.22
    if abs(x) > 0.28 and ath > 50 and y > clav - 0.05:
        # likely shoulder/arm — skip rewrite
        return None

    # -------- Pectorals (anterior mound) --------
    # Smooth inframammary: gentle curve, not deep W
    lat_n = min(1.0, abs(x) / 0.14)
    imf_curve = imf - 0.006 + 0.04 * (lat_n ** 1.6)
    top_pec = clav + 0.005
    # medial to sternum, lateral toward anterior axillary
    if (imf_curve - 0.01) <= y <= top_pec and ath <= 72:
        front = z - axz
        if front > -0.01:
            # widen mid-breast, taper near axilla
            max_ath = 52 + 18 * max(0.0, min(1.0, (y - imf_curve) / max(0.04, top_pec - imf_curve)))
            if ath <= max_ath and front > -0.005:
                if x < -0.004:
                    return "right_pectoral_region"
                if x > 0.004:
                    return "left_pectoral_region"
                return "right_pectoral_region" if x <= 0 else "left_pectoral_region"

    # -------- Abdomen (front, below pecs) --------
    ab_top = imf_curve - 0.008
    ab_top_center = imf - 0.018
    if ath < 30:
        ab_top = ab_top_center + (ab_top - ab_top_center) * (ath / 30.0)
    ab_bottom = iliac + 0.05
    if ab_bottom <= y <= ab_top and ath <= 58:
        front = z - axz
        if front > 0.005:
            return "full_abdomen_region"

    # -------- Ribs (lateral wrap) --------
    rib_top = clav - 0.04
    rib_bot = waist - 0.03
    if rib_bot <= y <= rib_top:
        if 48 <= ath <= 135:
            if ath >= 55 or y < imf_curve - 0.015:
                if x < 0:
                    return "right_ribs_region"
                if x > 0:
                    return "left_ribs_region"

    # -------- Upper back --------
    ub_top = neck - 0.01
    ub_bot = scap - 0.01
    if ub_bot <= y <= ub_top and ath >= 95:
        if abs(x) < 0.22:
            return "upper_back_region"

    # -------- Lower back --------
    lb_top = scap
    lb_bot = iliac + 0.035
    if lb_bot <= y <= lb_top and ath >= 90:
        if abs(x) < 0.20:
            return "lower_back_region"

    return None


def clear_torso_pixels(pixels, w, h, torso_colors, bg):
    """Set torso palette colors to background (exact match)."""
    tol = 3 / 255.0
    n = w * h
    for i in range(n):
        o = i * 4
        r, g, b = pixels[o], pixels[o + 1], pixels[o + 2]
        for cr, cg, cb, _ in torso_colors:
            if abs(r - cr) <= tol and abs(g - cg) <= tol and abs(b - cb) <= tol:
                pixels[o] = bg[0]
                pixels[o + 1] = bg[1]
                pixels[o + 2] = bg[2]
                pixels[o + 3] = 1.0
                break


def paint_mesh_to_uv(obj, img, palette, lm_data):
    w, h = img.size
    pixels = list(img.pixels)
    regions = palette["regions"]
    bg = hex_rgb(palette["background"]["authoringColor"])
    torso_colors = [hex_rgb(regions[rid]["authoringColor"]) for rid in TORSO_IDS]
    clear_torso_pixels(pixels, w, h, torso_colors, bg)

    _to_blender, to_runtime, mode = detect_runtime_to_blender(obj)
    print("AXIS_MODE", mode)

    levels = lm_data["levels"]
    axis_samples = lm_data.get("axisZSamples", [])
    mesh = obj.data
    mw = obj.matrix_world
    uv_layer = mesh.uv_layers.active.data

    color_of = {rid: hex_rgb(regions[rid]["authoringColor"]) for rid in TORSO_IDS}

    # Dense sample each triangle (barycentric grid)
    STEPS = 7
    painted = 0
    for poly in mesh.polygons:
        loops = list(poly.loop_indices)
        verts = list(poly.vertices)
        if len(loops) < 3:
            continue
        # triangulate fan
        for t in range(1, len(verts) - 1):
            tri_v = [verts[0], verts[t], verts[t + 1]]
            tri_l = [loops[0], loops[t], loops[t + 1]]
            p0 = mw @ mesh.vertices[tri_v[0]].co
            p1 = mw @ mesh.vertices[tri_v[1]].co
            p2 = mw @ mesh.vertices[tri_v[2]].co
            uv0 = uv_layer[tri_l[0]].uv
            uv1 = uv_layer[tri_l[1]].uv
            uv2 = uv_layer[tri_l[2]].uv
            for i in range(STEPS + 1):
                for j in range(STEPS + 1 - i):
                    a = i / STEPS
                    b = j / STEPS
                    c = 1.0 - a - b
                    if c < -1e-6:
                        continue
                    pos = p0 * c + p1 * a + p2 * b
                    uv = uv0 * c + uv1 * a + uv2 * b
                    p_rt = to_runtime(pos)
                    rid = classify_torso(p_rt, lm_data, levels, axis_samples)
                    if rid is None:
                        continue
                    col = color_of[rid]
                    # Match tools/body-mask bake + runtime sampler: image v flips mesh UV.y
                    x = int(max(0, min(w - 1, round(uv.x * (w - 1)))))
                    y = int(max(0, min(h - 1, round((1.0 - uv.y) * (h - 1)))))
                    # 3x3 stamp for UV coverage across seams
                    for dy in (-1, 0, 1):
                        for dx in (-1, 0, 1):
                            xx = max(0, min(w - 1, x + dx))
                            yy = max(0, min(h - 1, y + dy))
                            o = (yy * w + xx) * 4
                            pixels[o] = col[0]
                            pixels[o + 1] = col[1]
                            pixels[o + 2] = col[2]
                            pixels[o + 3] = 1.0
                            painted += 1

    img.pixels = pixels
    img.update()
    print("PAINTED_STAMPS", painted)
    return painted


def fill_sternum_seam(img, palette):
    """Ensure left/right pecs meet at center without tab — grow toward black gap near sternum band."""
    # Lightweight: already assigned by x sign; skip heavy morph.
    return


def main():
    BACKUPS.mkdir(parents=True, exist_ok=True)
    stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    if AUTHORING.exists():
        shutil.copy2(AUTHORING, BACKUPS / f"neutro_body_v1_anatomical_regions_pre_curate_{stamp}.png")

    palette = json.loads(PALETTE.read_text(encoding="utf-8"))
    lm_data = json.loads(LANDMARKS.read_text(encoding="utf-8"))

    obj = find_body()
    if obj is None:
        print("ERROR no body")
        sys.exit(1)
    if obj.name != PAINTABLE:
        obj.name = PAINTABLE

    img = load_image()
    paint_mesh_to_uv(obj, img, palette, lm_data)
    fill_sternum_seam(img, palette)

    img.filepath_raw = str(AUTHORING)
    img.file_format = "PNG"
    img.save()
    shutil.copy2(AUTHORING, BACKUPS / f"neutro_body_v1_anatomical_regions_curated_{stamp}.png")
    print("WROTE", AUTHORING)
    print("CURATE_TORSO_OK")


if __name__ == "__main__":
    main()
